### PR TITLE
fix: tolerate missing partition folders in 6.2.1 -> 6.3 migration

### PR DESCRIPTION
**Pull request by ideoma**
When last partition exists in `_txn` file but does not exist on disk (perhaps QuestDB was hard aborted) Mig620 fails.

The fix will tolerate missing partitions, including missing last partition assuming column presence in the missing partition is the same as in the proceeding partition.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 17 / 19 (89.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|17|19|89.47%|

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/test/java/io/questdb/test/tools/TestUtils.java [hunk 1]
```diff
@@ -513,6 +525,24 @@ public static long connect(long fd, long sockAddr) {
         return Net.connect(fd, sockAddr);
     }
 
+    public static void copyDirectory(Path from, Path to, int dirMode) throws IOException {
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/test/java/io/questdb/test/tools/TestUtils.java]: done in https://github.com/questdb/questdb/pull/2114/commits/1b73cf180b3bc9f9935cbc2e379f3069ba774059

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 29 / 32 (90.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|12|13|92.31%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 29 / 32 (90.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|12|13|92.31%|

**Review by marregui — DISMISSED** [review]
LGTM

File: core/src/test/java/io/questdb/test/tools/TestUtils.java [hunk 2]
```diff
@@ -46,9 +46,12 @@
 import org.junit.Assert;
 
 import java.io.*;
+import java.nio.file.FileSystems;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
```
Comment (marregui) [inline core/src/test/java/io/questdb/test/tools/TestUtils.java:53]: not used either :)

File: core/src/test/java/io/questdb/test/tools/TestUtils.java [hunk 3]
```diff
@@ -46,9 +46,12 @@
 import org.junit.Assert;
 
 import java.io.*;
+import java.nio.file.FileSystems;
```
Comment (ideoma) [inline core/src/test/java/io/questdb/test/tools/TestUtils.java:49]:
```suggestion
```

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 29 / 32 (90.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|12|13|92.31%|

File: core/src/test/java/io/questdb/test/tools/TestUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/test/java/io/questdb/test/tools/TestUtils.java]: done

File: core/src/test/java/io/questdb/test/tools/TestUtils.java (hunk 2 above)
Comment (ideoma) [reply core/src/test/java/io/questdb/test/tools/TestUtils.java]: done

**Review by marregui — APPROVED** [review]
LGTM

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 29 / 32 (90.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|17|19|89.47%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|12|13|92.31%|

**Pull request merged**